### ISSUE TITLE
Remove `ProfilerHparams` from `Trainer.__init__`

### ISSUE THREAD

**Issue by hanlint** [mosaicml/composer#306]
For ease of use, the init signature should never require someone to import a custom class -- must be python/pytorch native. The only exception is `model`, which has to inherit from `MosaicModel`.

**Comment by ravi-mosaicml** [comment]
Init only takes Python and PyTorch primitives, except for `model` of type `BaseComposerModel` and `load_object_store` of type `ObjectStoreProvider`. For users who are downloading a checkpoint from a private object store, I think it's OK for them to have to import a custom class. Closing for now.

**Comment by ravi-mosaicml** [comment]
Re-opening as the profiler still needs to be addressed. Renaming the issue for that.

**Comment by ravi-mosaicml** [comment]
A few options (summarized from the slack discussion):

1. (credit @hanlint):
```
profiler: Profiler
profiler_types: Sequence[Union[str, Callback??]]) # can be [TorchProfiler, DataloderProfiler, etc.], or simply 'torch', 'dataloader', 'system', 'all' which will create the default objects.
```

2. (credit: @hanlint)
```
profilers: Sequence[[Union[str, Callback]]] # 
profiler_config: Dict[str, Any] # basically used to create the `Profiler` object, we then default to using the JSON Event Handler
```

3.
* `skip_first` , `wait`, `warmup`, `active`, and `repeat` can be passed in directly like we did with the checkpointer
* The profiler trace event handler can be a `Sequence[ProfilerEventHandler]`. If not specified, then it will default to the json trace handler, which is the only one that we have implemented. It would be an advanced feature for a user to construct their own trace handler, so an import here is OK.
* The tricky one is the list of profilers (dataloader, system, and/or torch), since each has its own configuration options. For the list of profilers, I think we could go with a `Sequence[SubProfiler]` If not specified, it defaults to all 3 without any custom config. If a user wants custom config for any profiler, then they would need to add the imports.
